### PR TITLE
Add %% repetition

### PR DESCRIPTION
**Pull request by CAD97**
```
A* %% B => (A+ %% B)?
A+ %% B =>  A+ % B B?
```

This means that %% is a separated list with optional trailing separator, but does not allow a single lone comma.

------

Migration of https://github.com/lykenware/gll/pull/77 for more direct handling. Quoting myself from that thread:

> I have determined that it's not an issue with the gll_macros-macro-compiled version of the grammar (though that one for some reason doesn't allow whitespace?): https://github.com/CAD97/gll/blob/test-poc/tests/modmod.rs (run with `cargo test --package gll --test modmod main -- --exact`)
> 
> I am just more confused now, but the issue must be with the macro_rules bootstrap of the meta grammar. Somehow.

### CODE REVIEW

File: src/grammar.rs [hunk 1]
```diff
@@ -216,6 +237,27 @@ impl<Pat> RuleWithNamedFields<Pat> {
...
+                    paths
+                        .into_iter()
+                        .map(|mut path| {
+                            path.insert(0, 0);
+                            path
+                        })
+                        .collect(),
+                )
+            })
+            .collect();
```
Comment (Centril) [inline src/grammar.rs:256]: From the structure of things ISTM that everything up to here is identical between `repeat_{many|more}{_trailing}?` -- refactor?

File: src/generate/rust.rs [hunk 2]
```diff
@@ -143,6 +144,11 @@ impl<Pat: Ord + Hash + RustInputPat> Rc<Rule<Pat>> {
                 elem.parse_node_kind(parse_nodes).0,
                 sep.parse_node_kind(parse_nodes).0
             )),
+            Rule::RepeatManyTrailing(elem, sep) => ParseNodeKind(format!(
+                "({}* %% {})",
+                elem.parse_node_kind(parse_nodes).0,
+                sep.parse_node_kind(parse_nodes).0
+            )),
             Rule::RepeatMore(rule, None) => ParseNodeKind(format!(
                 // FIXME(rust-lang-nursery/rustfmt#3004) work around rustfmt removing trailing +.
```
Comment (Centril) [inline src/generate/rust.rs:171]: (issue seems closed)

File: build.rs [hunk 3]
```diff
@@ -106,10 +106,13 @@ fn main() {
...
             Modifier =
                 {Opt:"?"} |
-                {Repeat:{ {repeat:Repeat} {{ "%" {sep:Primary} }?} }};
+                {Repeat:{ {repeat:Repeat} {{ {kind:RepeatKind} {sep:Primary} }?} }};
             Repeat =
                 {Many:"*"} |
                 {More:"+"};
+            RepeatKind =
+                {One:"%"} |
+                {Two:"%%"};
```
Comment (eddyb) [inline build.rs]: Maybe this should be `SepKind` with `Simple` and `Trailing`?

File: macros/tests/basic.rs [hunk 4]
```diff
@@ -147,3 +147,10 @@ testcases![
     b: 1:3-1:4
 }"
 ];
+
+testcases! [
```
Comment (eddyb) [inline macros/tests/basic.rs]: Should just add to the one above.

File: macros/tests/basic.rs [hunk 5]
```diff
@@ -147,3 +147,10 @@ testcases![
...
+    repeat_many_trailing_0 {
+        A = "a"* %% "b";
+    }: A("") => "\
```
Comment (eddyb) [inline macros/tests/basic.rs]: Please use `"abab"` for testing (if we had multiple testcases per grammar, `"aba"` would also be nice to test).

File: src/grammar.rs [hunk 6]
```diff
@@ -319,7 +361,9 @@ pub enum Rule<Pat> {
 
     Opt(Rc<Rule<Pat>>),
     RepeatMany(Rc<Rule<Pat>>, Option<Rc<Rule<Pat>>>),
+    RepeatManyTrailing(Rc<Rule<Pat>>, Rc<Rule<Pat>>),
```
Comment (eddyb) [inline src/grammar.rs:364]: IMO the distinction should go in the `Option` in the cases above - either next to the `Rc<Rule<Pat>>` (in a tuple), or wrapping it.

You can use the suggested names here for an `enum`: https://github.com/rust-lang-nursery/gll/pull/2#discussion_r264797019.

File: src/generate/rust.rs (hunk 2 above)
Comment (eddyb) [reply src/generate/rust.rs:171]: Please open an issue on lykenware/gll for cleanup (or a PR on this repo to remove the ` HACK` from that string).

**Comment by CAD97** [comment]
and we're back to https://github.com/rust-lang/gll/pull/77

**Pull request closed without merging**